parquet-hadoop/src/main/java/org/apache/parquet/filter2/dictionarylevel/DictionaryFilter.java

```diff
@@ -0,0 +1,356 @@
+
+    for (int i=0; i<=dict.getMaxId(); i++) {
+      switch(meta.getType()) {
+        case BINARY: dictSet.add(dict.decodeToBinary(i));
+          break;
+        case INT32: dictSet.add(dict.decodeToInt(i));
+          break;
+        case INT64: dictSet.add(dict.decodeToLong(i));
+          break;
+        case FLOAT: dictSet.add(dict.decodeToFloat(i));
+          break;
+        case DOUBLE: dictSet.add(dict.decodeToDouble(i));
+          break;
+        default:
+          LOG.warn("Unknown dictionary type" + meta.getType());
+      }
+    }
+
+    return (Set<T>) dictSet;
+  }
+
+  @Override
+  public <T extends Comparable<T>> Boolean visit(Eq<T> eq) {
+    Column<T> filterColumn = eq.getColumn();
+    ColumnChunkMetaData meta = getColumnChunk(filterColumn.getColumnPath());
+
+    // if the chunk has non-dictionary pages, don't bother decoding the
+    // dictionary because the row group can't be eliminated.
+    if (hasNonDictionaryPages(meta)) {
+      return BLOCK_MIGHT_MATCH;
+    }
+
+    T value = eq.getValue();
+
+    filterColumn.getColumnPath();
+
+    try {
+      Set<T> dictSet = expandDictionary(meta);
+      if (dictSet != null && !dictSet.contains(value)) {
+        return BLOCK_CANNOT_MATCH;
+      }
+    } catch (IOException e) {
+      LOG.warn("Failed to process dictionary for filter evaluation.", e);
+    }
+
+    return BLOCK_MIGHT_MATCH; // cannot drop the row group based on this dictionary
+  }
+
+  @Override
+  public <T extends Comparable<T>> Boolean visit(NotEq<T> notEq) {
+    Column<T> filterColumn = notEq.getColumn();
+    ColumnChunkMetaData meta = getColumnChunk(filterColumn.getColumnPath());
+
+    // if the chunk has non-dictionary pages, don't bother decoding the
+    // dictionary because the row group can't be eliminated.
+    if (hasNonDictionaryPages(meta)) {
+      return BLOCK_MIGHT_MATCH;
+    }
+
+    T value = notEq.getValue();
+
+    filterColumn.getColumnPath();
+
+    try {
+      Set<T> dictSet = expandDictionary(meta);
+      if (dictSet != null && dictSet.size() == 1 && dictSet.contains(value)) {
+        return BLOCK_CANNOT_MATCH;
+      }
+    } catch (IOException e) {
+      LOG.warn("Failed to process dictionary for filter evaluation.", e);
+    }
+
+    return BLOCK_MIGHT_MATCH;
+  }
+
+  @Override
+  public <T extends Comparable<T>> Boolean visit(Lt<T> lt) {
+    Column<T> filterColumn = lt.getColumn();
+    ColumnChunkMetaData meta = getColumnChunk(filterColumn.getColumnPath());
+
+    // if the chunk has non-dictionary pages, don't bother decoding the
+    // dictionary because the row group can't be eliminated.
+    if (hasNonDictionaryPages(meta)) {
+      return BLOCK_MIGHT_MATCH;
+    }
+
+    T value = lt.getValue();
+
+    filterColumn.getColumnPath();
+
+    try {
+      Set<T> dictSet = expandDictionary(meta);
+      if (dictSet == null) {
+        return BLOCK_MIGHT_MATCH;
+      }
+
+      for(T entry : dictSet) {
+        if(value.compareTo(entry) > 0) {
```

Comment:
No, I think this is correct. The logic changes to > because the order is reversed, not because the logic is negated.

If `V` is the bound and we find any value, `x`, in the dictionary such that `x < V`, then there may be a matching row. But we call `V.compareTo(x)` and the order is reversed. We could call `x.compareTo(V)`, but it seems like calling the method on `V` is more likely to result in something the JVM can optimize.

The tests also validate that this behavior is correct by testing the boundary conditions with the smallest and largest values.